### ISSUE TITLE
Is it possible to use a certain Hardware address(MAC) when generating veths in cni?

### ISSUE THREAD

**Issue by randmonkey** [projectcalico/cni-plugin#340]
Since there is an issue in udev https://github.com/systemd/systemd/issues/3374, it is possible that cni plugin fails to create veth pairs, that causes failures in configuring network of containers. In this udev issue we see that we can set a MAC when creating veths to avoid this bug. So, I would like to try to set MAC for veths in CNI, as a possible solution of https://github.com/projectcalico/calico/issues/763. I would like to get some suggestions on feasiblity and implemetation of these things.  
## Expected Behavior

Hardware addresses(MAC) of vethes in containers are set by cni 
## Current Behavior

CNI does not set MAC when creating veths, leaving it to systemd/udevd.
## Possible Solution

generate a hardware address from IP allocation result, and use the MAC in LinkAttr when calling netlink.LinkAdd to add veths in func DoNetworking in utils/network.go.

## Context

My team deployed a k8s + calico cluster, and we saw that a few containers has no network. We checked network interfaces, and did not find the veths for those containers. Then, we checked system logs and confirmed the failure is caused by https://github.com/systemd/systemd/issues/3374. So, I got an idea to set a MAC in CNI to avoid this udev issue.
## Your Environment

* CNI plugin version (e.g. `/opt/cni/bin/calico --version`): 1.6.1 (behavior still the same in the newest release 1.8.3)
* Orchestrator version (e.g. kubernetes, mesos, rkt): kunernetes
* CNI config:
```
{
    "name": "k8s-pod-network",
    "type": "calico",
    "etcd_endpoints": "http://x.x.x.x:2379",
    "etcd_key_file": "",
    "etcd_cert_file": "",
    "etcd_ca_cert_file": "",
    "log_level": "info",
    "ipam": {
        "type": "calico-ipam"
    },
    "policy": {
        "type": "k8s",
        "k8s_api_root": "https://x.x.x.x:443",
        "k8s_auth_token": "================================="
    },
    "kubernetes": {
        "kubeconfig": "/etc/cni/net.d/calico-kubeconfig"
    }
}
```
* Operating System and version: Ubuntu 16.04.2 LTS (GNU/Linux 4.4.0-62-generic x86_64)
* Link to your project (optional):

## Possible concerns 
What if the CNI generated a MAC duplicating to other virtual interfaces? can we avoid that?

**Comment by tmjd** [comment]
@caseydavenport Is there any reason the CNI-plugin couldn't or shouldn't do this?  Seems like a reasonable thing to me.
We would certainly need to avoid creating any duplicate MAC addresses though I feel like that is probably doable to some degree at least on a single host.  Though it would cause problems if a duplicate was created that another host had that was reachable in the same L2 segment.  That could probably be alleviated some by checking arp entries.  Setting the Local bit in the MAC address may help with this.

**Comment by randmonkey** [comment]
To avoid duplicate MAC, I think we can use the same method as naming cali* interfaces that is to use some bits in container ID. For example, we can use the first 12 16-ary digits(6 bytes) of container ID changed by setting the Local bit as the MAC.

**Comment by fasaxc** [comment]
I can't see a reason not to do this.  The MAC address is pretty irrelevant in a Calico network since we do not bridge veths together.  That means the MAC is only used for the single hop to the other side of the veth.  In OpenStack deployments, I think we force all the MACs of the tap devices to be the same so that VM migration is simpler.

**Comment by randmonkey** [comment]
I want to talk about some details about implementation of this issue.
 - Should "Generate MAC by CNI" be a configurable option, and what is the default value of the option? (my opinion: should be configurable, and default to be `true` - MAC is generated by CNI by default.  )
 - How to generate this MAC address? (my idea: first 2 digit is a fixed value to mark that the address is internal and generated by CNI, e.g: `ca`; the following 10 digits comes from container ID of infrastructure container. For example, if the infrastructure container ID is `012345abcdef....`, the MAC is `ca:01:23:45:ab:cd` )

**Comment by caseydavenport** [comment]
This was fixed in #432
